### PR TITLE
Download platform/architecture specific binaries

### LINKED ISSUE #20: split package to reduce install size
Hi.

im using ffmpeg-static together with fluent-ffmpeg (https://github.com/fluent-ffmpeg/node-fluent-ffmpeg) and that is causing a rather strange issue, the executables for **all** platforms are included and thus the total installed size is huge, over 200mb huge.

### PR DESCRIPTION
**Pull request by jperl**
Resolves #20 
Resolves https://github.com/qawolf/qawolf/issues/316 

To Do

- [x] Create GitHub action to download & release binaries on v* tags
- [x] Create install.js to download binary from release
- [x] Use separate `binary_release` for the version
- [x] Clean up install.js
- [x] Use non-deprecated github action
- [x] Alert user in install.js if they have an invalid platform
- [x] Compress to reduce download size

### CODE REVIEW

**Comment by jperl** [comment]
Instead of rely on the package version I will create a separate binary version, and include it in the package.json. That way the release github action can be run/updated separately from the package version. I cannot think of another way to make the unit tests pass.

File: .github/workflows/release-binaries.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+name: Release Binaries
+
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Download Binaries
+        run: sh ./build/index.sh
+      - uses: meeDamian/github-release@1.0
```
Comment (derhuerst) [inline .github/workflows/release-binaries.yml]: This action has just been deprecated.

> NOTICE: This action is now deprecated, and official actions should be used instead:
>
> https://github.com/actions/create-release
> https://github.com/actions/upload-release-asset

File: .travis.yml [hunk 2]
```diff
@@ -1,14 +1,15 @@
 language: node_js
 node_js:
-  - 'stable'
-  - 'lts/*'
-  - 'lts/dubnium'
+  - "stable"
+  - "lts/*"
+  - "lts/dubnium"
+cache:
+  npm: false
```
Comment (derhuerst) [inline .travis.yml]: Why this?

Comment (jperl) [reply .travis.yml]: That is not necessary, initially I thought it was skipping the install script but it was not. Removed in last commit.

File: index.js [hunk 3]
```diff
@@ -1,24 +1,22 @@
-var os = require('os')
-var path = require('path')
+var os = require("os");
```
Comment (derhuerst) [inline index.js]: Please revert these changes. They make it really hard to find out later when & how something was changed.

File: install.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
+var fs = require("fs");
+var os = require("os");
+var URL = require("url");
+var pkg = require("./package");
+
+function httpRequest(url, method, response) {
```
Comment (derhuerst) [inline install.js]: Use a slightly-higher-level lib like [`simple-get`](https://www.npmjs.com/package/simple-get) for this. Unfortunately we'll still disrepect other important parts of the HTTP ecosystem, e.g. caching & proxy support.

**Comment by derhuerst** [comment]
> Now the binaries github action can be triggered and released separately from the package version.

I think that makes a lot of sense because it decouples ffmpeg releases from `ffmpeg-static` releases (in both ways).

> This makes it possible to run the tests against different binary releases. Otherwise you would not be able to run the tests until after a release.

The tests don't assert the functionality of the ffmpeg binaries anyways, so we don't need that.

**Comment by jperl** [comment]
@derhuerst Sorry for the delay on wrapping this up I am taking a vacation this week and working on this when I can. I believe the PR should be ready for review now.

The last commit tests are failing because I changed the download url to eugeneware and there is no release with binaries available.

**Comment by jperl** [comment]
After this PR is merged you should create a `b3.0.0` tag so a binary release is created. Any time you want to update which binary release is downloaded by default, you just change [`ffmpeg-static.binary_release`](https://github.com/eugeneware/ffmpeg-static/pull/33/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R17) in the package.json.

You can also override it when you install the package with [an environment variable](https://github.com/eugeneware/ffmpeg-static/pull/33/files#diff-f16acefe4b6553580c43edab685f50f3R64) ex `FFMPEG_BINARY_RELEASE=b4.0.0 npm install`

**Comment by jperl** [comment]
I released this as @qawolf/ffmpeg-static and used it in qawolf to check it works. The [tests](https://github.com/qawolf/qawolf/pull/325 pass) and the install works properly.

**Comment by jperl** [comment]
Leaving the url as qawolf so tests pass. Will change it back before merging. I updated it to compress the binaries so the download size is much smaller.

File: .travis.yml [hunk 5]
```diff
@@ -1,14 +1,13 @@
 language: node_js
 node_js:
-  - 'stable'
-  - 'lts/*'
-  - 'lts/dubnium'
+  - "stable"
```
Comment (derhuerst) [inline .travis.yml]: Please revert this change.

File: build/index.sh [hunk 6]
```diff
@@ -19,38 +19,53 @@ echo '  downloading from ffmpeg.zeranoe.com'
 download 'https://ffmpeg.zeranoe.com/builds/win64/static/ffmpeg-latest-win64-static.zip' win32-x64.zip
 echo '  extracting'
 unzip -o -d ../bin/win32/x64 -j win32-x64.zip '**/ffmpeg.exe'
+echo '  compressing'
+$tar_exec -C ../bin/win32/x64 -czvf ../bin/win32/x64/ffmpeg.tar.gz ffmpeg.exe
```
Comment (derhuerst) [inline build/index.sh]: Out of curiosity: Why compress the executable before uploading? Modern web servers & CDNs should do this on-the-fly anyways, right?

Also, keep in mind that [the `tar` package adds another 7 dependencies](http://npm.broofa.com/?q=tar) just needed for *installing* `ffmpeg-static`.

File: install.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+  const promise = new Promise((x, y) => {
+    fulfill = x;
+    reject = y;
+  });
+
+  get(url, function(err, response) {
+    if (err || response.statusCode !== 200) {
+      const error = new Error(`Download failed. URL: ${url}`);
+      reject(error);
```
Comment (derhuerst) [inline install.js]: To retain helpful debugging info, add the original `err` to `error`, or reject with `err` itself.

File: install.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      });
+    }
+  });
+
+  return promise;
+}
+
+let progressBar = null;
+let lastDownloadedBytes = 0;
+function onProgress(downloadedBytes, totalBytes) {
```
Comment (derhuerst) [inline install.js]: Just call `onProgress` with the delta (taken from `chunk.length`) instead of the total. This will remove unnecessary calculations forth and back.

File: test/index.js [hunk 9]
```diff
@@ -8,5 +8,12 @@ it('should find ffmpeg', function(t) {
 
   var stats = fs.statSync(ffmpegPath);
   t.ok(stats.isFile(ffmpegPath));
+
+  try {
+    fs.accessSync(ffmpegPath, fs.constants.X_OK)
+  } catch {
```
Comment (derhuerst) [inline test/index.js]: Nice addition! 👍 Please use `t.doesNotThrow` to make this assertion show up in the TAP output even if it doesn't fail.

File: build/index.sh (hunk 6 above)
Comment (jperl) [reply build/index.sh]: After an investigation it seems the compressed version is slower or about the same time to install (except for one likely fluke), so I will revert the compression change. https://app.circleci.com/github/jperl/test-ffmpeg-install-speed/pipelines

File: install.js (hunk 8 above)
Comment (jperl) [reply install.js]: I need total bytes to instantiate the progress bar.

**Comment by jperl** [comment]
@derhuerst Thanks for the review 🙏. I made the changes you requested except https://github.com/eugeneware/ffmpeg-static/pull/33#discussion_r371338406

File: install.js (hunk 8 above)
Comment (derhuerst) [reply install.js]: You can change the signature to `onProgress(deltaBytes, totalBytes)`. Then you don't have to do calculations in two places.

Comment (jperl) [reply install.js]: Got it 👍

**Comment by jperl** [comment]
@derhuerst made the deltaBytes change. It should be good to go 👍

File: install.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+if (ffmpegPath) {
+  downloadFile(getDownloadUrl(), ffmpegPath, onProgress).then(() => {
+    // make executable
+    fs.chmodSync(ffmpegPath, 0o755);
+  });
+} else {
+  console.error(
+    "ffmpeg-static install failed: No binary found for architecture"
+  );
```
Comment (derhuerst) [inline install.js]: Sorry, noticed one more thing: Please add `process.exit(1)` to indicate that the installation has not been successful.

File: install.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    process.env.FFMPEG_BINARY_RELEASE || pkg["ffmpeg-static"]["binary_release"];
+  var url = `https://github.com/qawolf/ffmpeg-static/releases/download/${release}/${platform}-${arch}`;
+  return url;
+}
+
+if (ffmpegPath) {
+  downloadFile(getDownloadUrl(), ffmpegPath, onProgress).then(() => {
+    // make executable
+    fs.chmodSync(ffmpegPath, 0o755);
+  });
```
Comment (derhuerst) [inline install.js]: `.catch` this promise, `console.error(err)` and `process.exit(1)`.

**Comment by jperl** [comment]
Good catches. Updated it 👍

**Comment by jperl** [comment]
@derhuerst to wrap this up

1. create a `b*` tag to start a release
2. change the [download url](https://github.com/eugeneware/ffmpeg-static/blob/master/install.js#L60) from qawolf to eugeneware
3. update the binary_release in [package.json](https://github.com/eugeneware/ffmpeg-static/blob/master/package.json#L17)
